src/v/cloud_topics/level_zero/frontend_reader/reader.cc

```diff
@@ -335,44 +369,63 @@ ss::future<> level_zero_log_reader_impl::materialize_batches(
             _current = state::end_of_stream_state;
             co_return;
         }
-
-        // Patch materialized record batches
         auto batches = std::move(mat_res.value());
-        for (size_t i = 0; i < batches.size(); i++) {
-            auto& data_hdr = batches.at(i).header();
-            auto size = data_hdr.size_bytes;
-            auto crc = data_hdr.crc;
-            data_hdr = to_materialize_headers.at(i);
-            data_hdr.type = model::record_batch_type::raft_data;
-            data_hdr.size_bytes = size;
-            data_hdr.crc = crc;
-            // Recalculate the header crc
-            data_hdr.header_crc = model::internal_header_only_crc(data_hdr);
-        }
-
         // Propagate batches to the record batch cache
         if (cache_enabled()) {
             for (const auto& b : batches) {
                 _ct_api->cache_put(_underlying->ntp(), b.copy());
             }
         }
-
-        _batches = std::move(batches);
+        // Merge our selected subset of unhydrated batches with the materialized
+        // batches, preserving control batches from the local log.
+        auto batches_it = batches.begin();
+        chunked_circular_buffer<model::record_batch> hydrated;
+        for (auto it = _unhydrated.begin(); it != unhydrated_it; ++it) {
+            model::record_batch batch = ss::visit(
+              it->data,
+              [&it, &batches_it](const cloud_topics::extent_meta&) {
+                  model::record_batch batch = std::move(*batches_it);
```

Comment:
@Lazin will the read pipeline silently skip batches if no longer in L0 or will it throw? We want it to throw or I need to add the check here

Reply:
Looks like it should return an error so this is safe. Might be worth the check still...

Reply:
I added a check the number of batches we get back is the same as the number of extents we pass in